src/Lean/Environment.lean

```diff
@@ -797,18 +798,28 @@ def finalizeImport (s : ImportState) (imports : Array Import) (opts : Options) (
       moduleData   := s.moduleData
     }
   }
-  let env ← setImportedEntries env s.moduleData
-  let env ← finalizePersistentExtensions env s.moduleData opts
+  env ← setImportedEntries env s.moduleData
+  if leakEnv then
+    /- When allowed to leak `env`, we eagerly mark it persistent here, which
+       avoids RC updates as well as costly MT markings when e.g. an interpreter
+       closure (wich contains the environment) is put in an `IO.Ref` like would
+       happen with user environment extensions loaded in the next step. -/
+    env := Runtime.markPersistent env
+  env ← finalizePersistentExtensions env s.moduleData opts
+  if leakEnv then
+    /- Also mark environment extension states. -/
+    env := Runtime.markPersistent env
```

Comment:
You are calling `markPersistent` twice. It is worth explaining. I am guessing that you are concerned some extension may mark the environment as `MT` at `finalizePersistentExtensions`.

Reply:
That could happen; mostly I just want to give a consistent guarantee about what is returned with `leakEnv`.